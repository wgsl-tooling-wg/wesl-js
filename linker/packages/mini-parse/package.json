{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "run-s build:main build:util build:minified build:brotli",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite -c minified.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs && ls -l dist/minified.cjs.br",
    "format": "prettier . --write",
    "lint": "eslint src",
    "organize": "organize-imports-cli tsconfig.json",
    "prepublishOnly": "run-s build",
    "test": "vitest",
    "test:once": "vitest run"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./minified": {
      "types": "./dist/index.d.ts",
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.2.0",
    "@typescript-eslint/parser": "^8.2.0",
    "@vitest/ui": "^2.0.5",
    "berry-pretty": "^0.0.5",
    "esbuild": "^0.23.1",
    "eslint": "^9.9.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.31.6",
    "typescript": "^5.5.4",
    "vite": "^5.4.2",
    "vite-plugin-dts": "^4.0.3",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.0.5"
  }
}
