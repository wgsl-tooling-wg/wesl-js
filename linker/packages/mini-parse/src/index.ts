export * from "./MatchingLexer.js";
export * from "./Parser.js";
export * from "./ParserCollect.js";
export * from "./ParserCombinator.js";
export * from "./ParserLogging.js";
export * from "./ParserToString.js";
export * from "./ParserTracing.js";
export * from "./Span.js";
export * from "./SrcMap.js";
export * from "./SrcMapBuilder.js";
export * from "./TokenMatcher.js";
export * from "./WrappedLog.js";
