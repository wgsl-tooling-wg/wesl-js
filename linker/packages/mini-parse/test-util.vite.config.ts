/// <reference types="vitest/config" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
import { resolve } from "path";

const config = baseViteConfig();
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

export default defineConfig(config);
