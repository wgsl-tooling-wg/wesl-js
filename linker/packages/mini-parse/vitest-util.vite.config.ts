/// <reference types="vitest/config" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
import { resolve } from "path";

const config = baseViteConfig();
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-vitest-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/vitest-util/index.ts")];
lib.fileName = "vitestUtil";

export default defineConfig(config);