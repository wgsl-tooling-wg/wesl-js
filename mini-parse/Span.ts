/**
 * An range, from start (inclusive) to end (exclusive).
 */
export type Span = readonly [number, number];
