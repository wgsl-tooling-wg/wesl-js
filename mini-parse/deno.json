{
  "name": "@wesl/mini-parse",
  "version": "0.1.0",
  "exports": {
    ".": "./mod.ts",
    "./test-util": "./test-util/index.ts"
  },
  "tasks": {
    "dev": "deno test --allow-read --watch"
  },
  "license": "MIT",
  "imports": {}
}