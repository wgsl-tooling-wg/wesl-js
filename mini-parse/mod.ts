export * from "./MatchingLexer.ts";
export * from "./Parser.ts";
export * from "./ParserCombinator.ts";
export * from "./ParserLogging.ts";
export * from "./ParserTracing.ts";
export * from "./SrcMap.ts";
export * from "./TokenMatcher.ts";
