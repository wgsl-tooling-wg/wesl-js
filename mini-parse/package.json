{
  "name": "mini-parse",
  "version": "0.6.0-rc4",
  "type": "module",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "vite build",
    "test": "vitest"
  },
  "exports": {
    ".": {
      "types": "./dist/mod.d.ts",
      "import": "./dist/mod.js"
    }
  }
}
