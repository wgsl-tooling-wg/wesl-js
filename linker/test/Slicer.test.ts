import { SrcMap } from "@wesl/mini-parse";
import { expect, test } from "vitest";
import { assertSnapshot } from "@std/testing/snapshot";
import { sliceReplace } from "../Slicer.ts";

test("slice middle", async (ctx) => {
  const src = "aaabbbc";
  const srcMap = sliceReplace(src, [{ start: 3, end: 6, replacement: "X" }]);
  const { dest, entries } = srcMap;
  expect(dest).toBe("aaaXc");
  await assertSnapshot(ctx, entries);
});

test("slice end", () => {
  const src = "aaabbb";
  const srcMap = sliceReplace(src, [{ start: 3, end: 6, replacement: "X" }]);
  expect(srcMap.dest).toBe("aaaX");
  validateDestCovered(srcMap);
});

test("slice beginning", () => {
  const src = "aaabbb";
  const srcMap = sliceReplace(src, [{ start: 0, end: 3, replacement: "X" }]);
  validateDestCovered(srcMap);
  expect(srcMap.dest).toBe("Xbbb");
});

test("slice multiple", () => {
  const src = "aaabbbc";
  const srcMap = sliceReplace(src, [
    { start: 3, end: 6, replacement: "B" },
    { start: 0, end: 3, replacement: "A" },
  ]);
  validateDestCovered(srcMap);
  expect(srcMap.dest).toBe("ABc");
  expect(srcMap.entries.length).toBe(3);
});

test("slice none", () => {
  const src = "aaabbbc";
  const srcMap = sliceReplace(src, []);
  validateDestCovered(srcMap);
  expect(srcMap.dest).toBe(src);
});

test("slice none with start and end", () => {
  const src = "aaabbbc";
  const srcMap = sliceReplace(src, [], 3, 6);
  validateDestCovered(srcMap);
  expect(srcMap.dest).toBe("bbb");
});

test("slice one with start and end", () => {
  const src = "aaabbbc";

  const slices = [{ start: 3, end: 6, replacement: "B" }];
  const srcMap = sliceReplace(src, slices, 2);
  validateDestCovered(srcMap);
  expect(srcMap.dest).toBe("aBc");
});

test("slice with empty replacement", () => {
  const src = "aaabbbc";

  const slices = [{ start: 3, end: 6, replacement: "" }];
  const srcMap = sliceReplace(src, slices);
  validateDestCovered(srcMap);
  expect(srcMap.dest).toBe("aaac");
});

/** verify that the srcMap covers every part of the destination text */
function validateDestCovered(srcMap: SrcMap): void {
  const { dest, entries } = srcMap;
  let destPos = 0;
  entries.forEach(e => {
    expect(e.destStart).toBe(destPos);
    destPos = e.destEnd;
  });
  expect(destPos).toBe(dest.length);
}
