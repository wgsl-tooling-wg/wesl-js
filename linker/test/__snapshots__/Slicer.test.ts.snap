export const snapshot = {};

snapshot[`slice middle 1`] = `
[
  {
    destEnd: 3,
    destStart: 0,
    src: "aaabbbc",
    srcEnd: 3,
    srcStart: 0,
  },
  {
    destEnd: 4,
    destStart: 3,
    src: "aaabbbc",
    srcEnd: 6,
    srcStart: 3,
  },
  {
    destEnd: 5,
    destStart: 4,
    src: "aaabbbc",
    srcEnd: 7,
    srcStart: 6,
  },
]
`;
